JAVA-2111: Add HierarchyScanStrategy to configure inheritance behavior

tolbertam opened the pull request:
For [JAVA-2111](https://datastax-oss.atlassian.net/browse/JAVA-2111)

TODO:

- [x] validate that this actually works beyond unit level :)
- [x] Check the `scanAnnotations` field
- [x] Implement in Dao code once #1246 is merged
- [x] Documentation
- [x] Test all remaining DAO query methods

---

Motivation:

The ability to define properties and annotations of Entity-annotated
classes in parent classes or interfaces will make it easier for
implementors to reuse code and enforce behavior across multiple
entities.

In addition, the same applies for Dao-annotated interfaces.  For
example, being able to express a NullSavingStrategy on a shared parent
interface would be desirable for those who wish to share that
configuration among multiple Daos.

Modifications:

Added `@HierarchyScanStrategy` that can be defined anywhere in a Dao or
Entity type hierarchy.  This strategy dictates how to treat annotations
and properties defined in parent classes and interfaces.

By default, all parent interfaces and classes are considered.

Result:

The mapper can now make use of properties and annotations defined in
parent classes and interfaces of Dao and Entity-annotated types.

tolbertam committed:
JAVA-2111: Add HierarchyScanStrategy to configure inheritance behavior

Motivation:

The ability to define properties and annotations of Entity-annotated
classes in parent classes or interfaces will make it easier for
implementors to reuse code and enforce behavior across multiple
entities.

In addition, the same applies for Dao-annotated interfaces.  For
example, being able to express a NullSavingStrategy on a shared parent
interface would be desirable for those who wish to share that
configuration among multiple Daos.

Modifications:

Added `@HierarchyScanStrategy` that can be defined anywhere in a Dao or
Entity type hierarchy.  This strategy dictates how to treat annotations
and properties defined in parent classes and interfaces.

By default, all parent interfaces and classes are considered.

Result:

The mapper can now make use of properties and annotations defined in
parent classes and interfaces of Dao and Entity-annotated types.

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java:
```diff
@@ -0,0 +1,191 @@
+      TypeElement classElement, Types types) {
+    // Use the default HierarchyScanStrategy to find the configured HierarchyScanStrategy.
+    // This is done because the default strategy is the most permissive.
+    HierarchyScanStrategy strategy =
```
possible optimization could be done here such that if the strategy is not defined, return the fully traversed hierarchy so we don't have to traverse it again...but it's probably such a cheap operation that is only done for each dao/entity that it might not be worth it.

tolbertam replied:
Decided it's not really worth it, would require changing the API a bit and at this point things work great...

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -72,97 +77,107 @@ public DefaultEntityFactory(ProcessorContext context) {
-                  clusteringColumnIndex,
-                  previous.getGetterName(),
-                  property.getGetterName());
+        VariableElement field = findField(typeHierarchy, propertyName, typeMirror);
```
Might need some logic here to bound the highest type parsed for the field to where the getMethod was defined...not sure if that's worth it or not though.

olim7t replied:
It seems unlikely that a field with the same name would exist at a higher level _and_ be annotated,  I think we can safely ignore that case.

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -72,97 +77,107 @@ public DefaultEntityFactory(ProcessorContext context) {
-    CqlNameGenerator cqlNameGenerator = buildCqlNameGenerator(classElement);
-    Set<String> transientProperties = getTransientPropertyNames(classElement);
-
+    Set<String> encounteredPropertyNames = Sets.newTreeSet();
```
Why do we need `TreeSet`? Is ordering of property names important here?

tolbertam replied:
definitely not necessary, HashSet should suffice.  Will update.

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -179,15 +194,21 @@ public EntityDefinition getDefinition(TypeElement classElement) {
+          continue;
+        }
+        VariableElement field = (VariableElement) child;
+        if (field.getSimpleName().toString().equals(propertyName)
```
is `getSimpleName` check enough? Isn't there a risk that the same name of the property can be defined for types in different packages? Shouldn't we use ful qualified name?

tolbertam replied:
I think `getSimpleName` is what we want, and not sure there is a difference between the qualified and simple name for fields and if there was, simple name is probably preferred since we can scan multiple class for fields.   This is not something I changed (just shows up as a change from nesting things further in code).

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -179,15 +194,21 @@ public EntityDefinition getDefinition(TypeElement classElement) {
+        }
       }
     }
     return null;
```
Maybe we should follow a pattern of returning Optional instead of null? (similar to other annotation processor methods)

tolbertam replied:
this would mean passing optionals between methods, which is kinda ugly, but I see your point.  I agree that it's not good to be consistent though.  I'll change to standardize on not using Optional and using `@Nullable` where applicable.

tolbertam replied:
actually looking at this more, the two places we use Optional, we pass in as constructors to `DefaultEntityDefinition`.  I'm going to leave that be since it isn't relevant for this PR.   I will annotate arguments as nullable though where applicable.

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -364,11 +391,12 @@ private boolean isTransient(
 
-  private Set<String> getTransientPropertyNames(TypeElement classElement) {
-    TransientProperties transientProperties = classElement.getAnnotation(TransientProperties.class);
+  private Set<String> getTransientPropertyNames(Set<TypeElement> typeHierarchy) {
```
We will need to implement similar logic with a call to `getClassAnnotation` for every annotation that should be inherited, right?

tolbertam replied:
That's right, for now i've only done this for Entity as there's no Dao interface-level annotations yet (that I know of).

There may be an exception or two, such as the class-level `@CqlName` on an entity, which I think we would not want to inherit.

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/AnnotationScanner.java:
```diff
@@ -0,0 +1,74 @@
+        Set<Modifier> modifiers = child.getModifiers();
+        if (child.getKind() != ElementKind.METHOD
+            || modifiers.contains(Modifier.STATIC)
+            || modifiers.contains(Modifier.PRIVATE)) {
```
So if someone by mistake annotates private method we will just ignore that? Maybe some kind of log stating that we are not taking private methods into accounts?

tolbertam replied:
I think it's fine to quietly ignore these as we did the same with the previous mapper, we do the same for identifying property methods.

In addition, I don't think it can reasonably happen, you can't override a parent method that is private.  We could possibly even remove the check for static & private, but I think it's fine to be paranoid.

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -364,11 +391,12 @@ private boolean isTransient(
-    return transientProperties != null
-        ? Sets.newHashSet(transientProperties.value())
+    return annotation.isPresent()
+        ? Sets.newHashSet(annotation.get().getAnnotation().value())
```
Isn't that Set has always one element? Why do we need to return set here?

tolbertam replied:
the annotation value is an array, so there can be multiple elements.

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -436,20 +463,23 @@ private void scanFieldAnnotations(
+      Optional<? extends ResolvedAnnotation<? extends Annotation>> annotation =
+          getMethodAnnotation(annotationClass, getMethod, typeHierarchy);
+      if (annotation.isPresent()) {
         if (EXCLUSIVE_PROPERTY_ANNOTATIONS.contains(annotationClass)) {
```
what if someone will annotate a method with 3 exclusive annotations?
We will report in the `reportMultipleAnnotationError` method only the first two, right?
Maybe we should report all of them?

tolbertam replied:
I agree that it would be a nice to have, but i see it as low value for the effort it would take.  If have some time I'll look into it.

tomekl007 commented on mapper-processor/src/test/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScannerTest.java:
```diff
@@ -0,0 +1,208 @@
+    // when no HierarchyScanStrategy is defined, then the default behavior should be used
+    // which is to scan the entire tree and return it in the correct (bottom up) order.
+    assertThat(HierarchyScanner.resolveTypeHierarchy(c.classElement, types).toString())
+        .isEqualTo("[c, b, x, w, a, y, z]");
```
is the ordering of `a, y, z` is not important? Shouldn't it be `y,z,a`?

tolbertam replied:
a occurs before y and z in this context because b implements a.

If you look at the comment above with the graph the order of elements traversed should match the graph reading from bottom up, and left to right within the same depth of the graph, so for this case the order is encountered as

c
b x w
a y
z

since a has already been encountered, it's not added again.

tomekl007 commented on mapper-processor/src/test/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScannerTest.java:
```diff
@@ -0,0 +1,208 @@
+    MockInterface x = i("x", y);
+    MockInterface w = i("w");
+
+    MockClass b = c(this.getClass().getName(), null, a, r);
```
is there a particular reason why in this mock we choose to use `this.getClass().getName()` instead of the following pattern from different c mocks? (like `c("b", null, a,r)`
Same question to line https://github.com/datastax/java-driver/pull/1250/files#diff-26867dfb0fa8ecefa97f1251414342dcR85

tolbertam replied:
We need to do it this way because `HierarchyScanStrategy.highestAncestor` takes a `Class` and you cannot mock classes since they are final.

tolbertam replied:
with your suggestion to use `Classes.isSame()` I no longer need to do this :+1:

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java:
```diff
@@ -0,0 +1,191 @@
+public class HierarchyScanner {
+
+  // placeholder class for resolving the default HierarchyScanStrategy
+  @HierarchyScanStrategy
```
maybe we should specify values that we expect here explicitly instead of relying on defaults from `@HierarchyScanStrategy`?

tolbertam replied:
I think it's safe to assume the annotation defaults are the most permissive.  In addition this is also used a placeholder if no `@HierarchyScanStrategy` is present, so should match the default anyways.

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java:
```diff
@@ -0,0 +1,191 @@
+      if (superClass.getKind() == TypeKind.DECLARED) {
+        superClassElement = (TypeElement) types.asElement(superClass);
+        atHighestClass =
+            superClassElement
```
I think that you can simplify that check with the usage of `ProcessorContext.Classes.isSame()`

tolbertam replied:
nice, that's way better, and more correct. will use that.

tolbertam committed:
Process feedback and take care of some TODOs

* Feedback: use HashSet instead of TreeSet for encounteredPropertyNames.
* Feedback: use `@Nullable` on methods returning null and for
  parameters.
* Fix TODO: Consider highestAncestor for interfaces.
* Fix TODO: Support scanAncestors == false
* Fix TODO: Javadoc for HierarchyScanStrategy

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java:
```diff
@@ -0,0 +1,191 @@
+        classElement,
+        types,
+        (TypeElement t) -> {
+          HierarchyScanStrategy discoveredStrategy = t.getAnnotation(HierarchyScanStrategy.class);
```
I'm surprised that this works. Generally you can't inspect an annotation's `Class<?>` attribute during a processing round, because it might not be possible to load the referenced class yet (see `DefaultEntityFactory#readCustomConverterClasses` for example).

Have you tried with `HierarchyScanStrategy#highestAncestor()` being an application class?

tolbertam replied:
I'm not confident it will work as I haven't tested it yet at the compile phase, will see how it goes today ;)

tolbertam replied:
You're right, it doesn't work, will follow the strategy you referenced :+1:

```
Error:(676, 3) java: Unexpected error while writing generated code: javax.lang.model.type.MirroredTypeException: Attempt to access Class object for TypeMirror com.datastax.oss.driver.mapper.EntityPolymorphismIT.Asset
...
com.datastax.oss.driver.internal.mapper.processor.util.HierarchyScanner.traverseHierarchy(HierarchyScanner.java:136)
  	at com.datastax.oss.driver.internal.mapper.processor.util.HierarchyScanner.traverseFullHierarchy(HierarchyScanner.java:111)
  	at com.datastax.oss.driver.internal.mapper.processor.util.HierarchyScanner.resolveTypeHierarchy(HierarchyScanner.java:75)
```

tolbertam replied:
Fixed this in 8fdcc880a1f6050cf7b915127e29f5c4c1a56c25

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java:
```diff
@@ -0,0 +1,191 @@
+   *   <li>If <code>p != null</code> </code>Set <code>e := p, i := ni</code> and repeat starting at
+   *       step 3
+   *   <li>Visit <code>ni</code> interfaces and their parents until we've reached root
+   * </ol>
```
These explanations would be a bit easier to follow if you named your variables the same as in the code: `e` => `typeElement`, `i` => `interfacesToScan`, etc.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -72,97 +77,107 @@ public DefaultEntityFactory(ProcessorContext context) {
+        // skip properties we've already encountered.
+        if (encounteredPropertyNames.contains(propertyName)) {
+          continue;
+        }
```
If I read `HierarchyScanner` correctly, `typeHierarchy` is in ascending order, so we'll run into child classes first. This could be a problem if we want our annotations to be inherited, for example:
```java
interface Identifiable {
  @PrimaryKey UUID getId();
  void setId(UUID id);
}
@Entity
class Product implements Identifiable {
  @Override public UUID getId() { ... }
  @Override public void setId(UUID id) { ... }

}
```
Maybe making our annotations `@Inherited` would be enough to address this, but it's probably not resolved when parsing annotation mirrors.

Another approach could be to traverse the set in reverse order. But that would prevent a class from overriding a method with additional annotations.

In any case we need to define rules for annotation inheritance, write them down in the manual and test them.

olim7t replied:
Actually I just realized you're handling inheritance in `AnnotationScanner#getMethodAnnotation`, correct?

tolbertam replied:
That's right, it works similarly to the 3.x mapper, parsing out the properties is done separately from parsing the annotations.  Once you've identified the properties, you find the annotations, which can be pulled from different levels of the hierarchy.

tolbertam replied:
Although that is only the case for method annotations.  We only pull annotations from the identified field (also similar to 3.x mapper)

olim7t replied:
Sounds good, fields can't be overridden anyway so it's unlikely that the same one could be present at different levels.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -72,97 +77,107 @@ public DefaultEntityFactory(ProcessorContext context) {
+        String setMethodName = getMethodName.replaceFirst("get", "set");
+        ExecutableElement setMethod = findSetMethod(typeElement, setMethodName, typeMirror);
+        if (setMethod == null) {
+          continue; // must have both
```
This won't detect a property if the getter and setter come from different types in the hierarchy. That might be a fair requirement, but on the other hand it's pretty straightforward to scan the whole hierarchy for the setter, like you did for the field.

tolbertam replied:
Sure, I can add that.   I figured that `get` and `set` would be defined on the same type, but it's fair to define a super interface that only has getters.

tolbertam committed:
Add integration tests, Dao method inheritance support, feedback

Also fix some issues identified in test, such as identifying the correct
parameter and return types based on inheritance details.

tolbertam committed:
Add HierarchyScanStrategy integration tests, fix bugs

Fix issues around Dao inheritance where there are multiple parent
interfaces with generic types.

Also fix issues around HierarchyScanStrategy (use annotation mirrors)

Adjust mocking in HierarchyScanner, which is starting to get incredibly
fragile.

tolbertam committed:
Add inheritance annotation scanning in Dao for DefaultNullSavingStrategy

tolbertam committed:
SET_TO_NULL in Polymorphism test for 2.1 support

tolbertam committed:
Add documentation

tolbertam commented:
Added documentation, I think there still might be issues with generic type variables in some Dao query methods..working on testing and fixing that.

tolbertam committed:
Update GetEntity and SetEntity to work with typevar, tests

tolbertam committed:
Handle generic return types properly, add Query tests, fix regression

* Handle generic return types from Dao interface methods like
  CompletableFuture<T>.
* Add tests for `@Query`
* Fix regression in DaoGetEntityMethodGenerator

tolbertam committed:
Fix regression in SetEntity

tolbertam commented on manual/mapper/daos/README.md:
```diff
@@ -48,4 +48,68 @@ ProductDao dao = inventoryMapper.productDao("someKeyspace");
 
 The returned object is thread-safe, and can securely be shared throughout your application.
 
+### Inheritance
```
I considered making a dedicated page for entity/dao inheritance, but wasn't sure if it'd be worth it or not, instead I added a section for each entity & dao.

tolbertam commented on mapper-processor/src/test/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScannerTest.java:
```diff
@@ -0,0 +1,342 @@
+    final Name qfName;
+
+    @SuppressWarnings("unchecked")
+    MockElement(
```
The amount of mocking needed ended up getting out of control, perhaps we could consider removing this test now given we have decent integration test coverage.

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/generation/GeneratedCodePatterns.java:
```diff
@@ -63,18 +70,48 @@
     return result;
   }
 
+  private static TypeName getTypeName(TypeMirror mirror, Map<Name, TypeElement> typeParameters) {
```
Feels like there should be an easier way to do this, but was the best I could come up with.

tolbertam commented:
Think I have this all sorted out.  There were a ton of fun little edge cases that I had to work out, hopefully I caught them all!

tomekl007 commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/NullSavingStrategyIT.java:
```diff
@@ -171,6 +184,10 @@ ProductSimpleDaoGlobalLevelDoNotSetOverrideSetToNull productDaoLocalOverride(
 
+  @Dao
+  public interface ProductSimpleDaoSetToNullFromParentInterface
+      extends ProductSimpleDaoImplicit, SetToNull {}
```
Why do we need to extend `ProductSimpleDaoImplicit` (instead of inlining it in the `ProductSimpleDaoSetToNullFromParentInterface`)?
It seems that we are testing two aspects - maybe it will be better to split it into two tests?
Or maybe we should also test a case where ProductSimpleDaoImplicit should extend `SetToNull`?

tolbertam replied:
I chose to use `ProductSimpleDaoImplicit` because it provides methods we can use and isn't already annotated with `DefaultNullSavingStrategy`.  So this is to verify that a parent annotation is inherited.  We could have redefined the `ProductSimpleDaoImplicit`s methods here, but this seemed easier.

tomekl007 commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/NullSavingStrategyIT.java:
```diff
@@ -125,11 +131,18 @@ public void should_do_not_throw_when_construct_dao_with_global_level_SET_TO_NULL
         @DaoKeyspace CqlIdentifier keyspace);
   }
 
+  @DefaultNullSavingStrategy(NullSavingStrategy.SET_TO_NULL)
```
Will this work if the @`Mapper` will be annotated with `@DefaultNullSavingStrategy`?
If so maybe we should add a test for it.

tolbertam replied:
I don't believe so, at least I didn't intend for it to work that way.  We'd have to make `Mapper` annotations accessible for the Dao code, which is probably doable, but I think `Dao`-level and Daos supporting inheriting annotations may be powerful enough.

tomekl007 commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/EntityPolymorphismIT.java:
```diff
@@ -85,25 +89,40 @@ public static void setup() {
 
   // another parent interface with dao methods
-  interface WriteTimeDao<Y extends WriteTimeProvider> extends ShapeDao<Y> {
+  interface WriteTimeDao<Y extends WriteTimeProvider> extends BaseDao<Y> {
```
awesome

tomekl007 commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java:
```diff
@@ -0,0 +1,241 @@
+
+  private static class HierarchyScanStrategyOptions {
+    boolean scanAncestors = defaultStrategy.scanAncestors();
+    TypeMirror highestAncestor = null;
```
why we don't take `defaultStrategy.highestAncestor`?

tolbertam replied:
Good catch.  I think it ultimately doesn't have an effect, but we can get a TypeMirror for that (Object.class) :+1:

olim7t commented on manual/mapper/daos/README.md:
```diff
@@ -48,4 +48,68 @@ ProductDao dao = inventoryMapper.productDao("someKeyspace");
+[@Dao].
+
+In addition to inheriting methods from parent interfaces, interface-level annotations such as
+[DefaultNullSavingStrategy] are also inherited:
```
```suggestion
[@DefaultNullSavingStrategy] are also inherited:
```

olim7t commented on manual/mapper/daos/README.md:
```diff
@@ -48,4 +48,68 @@ ProductDao dao = inventoryMapper.productDao("someKeyspace");
+interface BaseDao {
+}
+
+interface RectangleDao extends BaseDao {};
```
Nit: should probably be annotated with `@Dao` (not relevant here but more realistic)

tolbertam replied:
thats an oversight, will fix.

olim7t commented on manual/mapper/daos/README.md:
````diff
@@ -48,4 +48,68 @@ ProductDao dao = inventoryMapper.productDao("someKeyspace");
+```
+
+Scanning priority is driven by proximity to the [@Dao] interface and within the same level 
+interfaces are given priority by the order they are in the type signature.  To control how the 
````
I find the wording a bit hard to understand.

tolbertam replied:
will clean up.

tolbertam replied:
cleaned this up a bit and referenced parts of the example where this applies.

olim7t commented on manual/mapper/entities/README.md:
````diff
@@ -303,6 +303,224 @@ If you want the name to be case-sensitive, it must be enclosed in double-quotes,
+
+Here is an example of a polymorphic mapping:
+
+```java
````
That's a long example. For the sake of conciseness, I think you could:
* omit all constructors
* put trivial methods on one line:

    ```java
    public UUID getId() { return id; }
    ```

tolbertam replied:
sounds good.   I was aiming to provide an example that would compile (so i could test that I didn't miss anything), I'll clean it up.

tolbertam replied:
Went ahead and updated the manual section for this. :+1:

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/CodeGeneratorFactory.java:
```diff
@@ -75,5 +77,7 @@
-      ExecutableElement methodElement, DaoImplementationSharedCode enclosingClass);
+      ExecutableElement methodElement,
+      Map<Name, TypeElement> typeParameters,
+      DaoImplementationSharedCode enclosingClass);
```
Nit: the javadoc of `newDaoImplementation` references this method, the link is now broken.

tolbertam replied:
fixed

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java:
```diff
@@ -73,11 +79,30 @@ protected TypeElement typeArgumentAsEntityElement(TypeMirror mirror) {
     return asEntityElement(declaredType.getTypeArguments().get(0));
   }
 
+  /** Evaluates whether or not this type argument could by an entity (declared or a typevar). */
```
```suggestion
  /** Evaluates whether or not this type argument could be an entity (declared or a typevar). */
```

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoGetEntityMethodGenerator.java:
```diff
@@ -93,8 +96,12 @@ public DaoGetEntityMethodGenerator(
     TypeMirror returnType = methodElement.getReturnType();
-    if (returnType.getKind() == TypeKind.DECLARED) {
-      Element element = ((DeclaredType) returnType).asElement();
+    if (couldBeEntity(returnType)) {
```
`couldBeEntity` is a bit of a deceptive name, because it also matches iterables of entities. What it really is is `isDeclaredOrTypeVar`. Or we could just inline it, it's only used in two places.

tolbertam replied:
I initially factored it into a separate method as was worried about missing a case and thought this may make it easier to communicate intent, but since it ended up only being needed in two places i'll change it back.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoGetEntityMethodGenerator.java:
```diff
@@ -93,8 +96,12 @@ public DaoGetEntityMethodGenerator(
+      if (element == null && returnType.getKind() == TypeKind.DECLARED) {
+        // if it's not an entity, treat it as a normal declared type.
+        element = ((DeclaredType) returnType).asElement();
+      }
```
This really confuses me. `asEntityElement` already checks the element kind, but we check it again at line 106. Also, line 106 produces a null warning on `entity`.

How about doing it this way instead:
```java
Transformation transformation = null;
TypeMirror returnType = methodElement.getReturnType();
TypeElement entityElement = asEntityElement(returnType);
if (entityElement != null) {
  entityElement = (TypeElement) element;
  transformation = parameterIsGettable ? Transformation.NONE : Transformation.ONE;  
} else if (returnType.getKind() == TypeKind.DECLARED) {
  Element element = ((DeclaredType) returnType).asElement();
  if (context.getClassUtils().isSame(element, PagingIterable.class)) {
    ... // no change
  } else if (context.getClassUtils().isSame(element, MappedAsyncPagingIterable.class)) {
    ... // no change
  }
}
```
```

tolbertam replied:
thats definitely more comprehensible :+1:

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/EntityPolymorphismIT.java:
```diff
@@ -0,0 +1,972 @@
+    Function<CqlIdentifier, SquareDao> squareDao = keyspace -> mapper.squareDao(keyspace);
+    Function<CqlIdentifier, SphereDao> sphereDao = keyspace -> mapper.sphereDao(keyspace);
+    return new Object[][] {
+      new Object[] {
```
```suggestion
      {
```
You don't need to repeat that for each inner array.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:
```diff
@@ -70,13 +79,45 @@
+  private final Map<Class<? extends Annotation>, Annotation> annotations;
+
+  private static final Set<Class<? extends Annotation>> ANNOTATIONS =
+      ImmutableSet.of(DefaultNullSavingStrategy.class);
```
That doesn't really bother me, but this has the same name as the field. So maybe `ANNOTATIONS_TO_SCAN` instead?

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:
```diff
@@ -70,13 +79,45 @@
+  }
+
+  @SuppressWarnings("unchecked")
+  private <A> Optional<A> getAnnotation(Class<A> annotationClass) {
```
```suggestion
  private <A extends Annotation> Optional<A> getAnnotation(Class<A> annotationClass) {
```
Avoids the "suspicious call" warning on `annotations.get`.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:
```diff
@@ -70,13 +79,45 @@
+
+  @SuppressWarnings("unchecked")
+  private <A> Optional<A> getAnnotation(Class<A> annotationClass) {
+    return Optional.ofNullable((A) annotations.get(annotationClass));
```
```suggestion
    return Optional.ofNullable(annotationClass.cast(annotations.get(annotationClass)));
```
Essentially the same, but it eliminates the need for the `@SuppressWarnings` on the method.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:
```diff
@@ -127,15 +168,80 @@ public String addQueryProvider(
     return implementationName;
   }
 
+  private Map<Name, TypeElement> parseTypeParameters(TypeMirror mirror) {
```
Since this is a pretty hairy method, maybe you could add comments that explain each part based on a common example.

E.g. `CircleDao extends ShapeDao<Circle>`, we are going through the method for `ShapeDao`.
* `mirror` is `ShapeDao<Circle>`
* `element` is `interface ShapeDao<T>`
* etc.

tolbertam replied:
I took a stab at explaining this, first in the javadoc for the method to explain an example of an interface and how it and its parents are processed when parsing type arguments, and inline in the code, hope that makes it more clear.

olim7t replied:
LGTM. To improve readability I would use a plain implementation comment for the method as well, we don't really need a javadoc here.

tolbertam replied:
:+1: will do

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:
```diff
@@ -127,15 +168,80 @@ public String addQueryProvider(
+      TypeElement element = (TypeElement) context.getTypeUtils().asElement(mirror);
+      DeclaredType declaredType = (DeclaredType) mirror;
+      for (int i = 0;
+          i < declaredType.getTypeArguments().size() && i < element.getTypeParameters().size();
```
Can they ever be different? (or is it to handle code that doesn't even compile)

tolbertam replied:
Ah nope, they can't. I previously had a bug (never committed) where I was pulling from the wrong element so I added this protection, but its no longer needed.

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/EntityPolymorphismIT.java:
```diff
@@ -0,0 +1,972 @@
+  interface WriteTimeDao<Y extends WriteTimeProvider> extends BaseDao<Y> {
+    @Insert(timestamp = ":writeTime")
+    void saveWithTime(Y y, long writeTime);
+  };
```
```suggestion
  }
```

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/EntityPolymorphismIT.java:
```diff
@@ -0,0 +1,972 @@
+  }
+
+  @Dao
+  interface SquareDao extends WriteTimeDao<Square>, ArbitraryInterface<Float> {};
```
```suggestion
  interface SquareDao extends WriteTimeDao<Square>, ArbitraryInterface<Float> {}
```

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/EntityPolymorphismIT.java:
```diff
@@ -0,0 +1,972 @@
+  interface CircleDao extends WriteTimeDao<Circle> {}
+
+  @Dao
+  interface SphereDao extends WriteTimeDao<Sphere> {};
```
```suggestion
  interface SphereDao extends WriteTimeDao<Sphere> {}
```

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:
```diff
@@ -127,15 +168,80 @@ public String addQueryProvider(
+                    element,
+                    "Could not resolve type parameter %s "
+                        + "on %s from child interfaces. This error usually means an interface "
+                        + "was inappropriately annotated with @Dao. Interfaces should only be annotated "
```
Nit: could you inject `Dao.class.getSimpleName()` instead of hardcoding it in the string?

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:
```diff
@@ -127,15 +168,80 @@ public String addQueryProvider(
+          // from the previously encountered interface and alias it to the
+          // type on this class.
+          TypeVariable genericType = (TypeVariable) typeArgument;
+          Name previousName = genericType.asElement().getSimpleName();
```
Sorry for being picky, but the names are throwing me off a bit here:
* why `genericType` for a type argument? (maybe inline it)
* `previousName`: if we're processing `BaseDao<T>` this would be `T` right? So why "previous"?

tolbertam replied:
Not a problem, agree that the naming is a bit confusing there, i'll clean it up.

tolbertam replied:
`previousName` definitely not right.   I'll change it to `typeVariableName` as thats more descriptive of what it is.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -73,97 +79,112 @@ public DefaultEntityFactory(ProcessorContext context) {
+                    PartitionKey.class.getSimpleName(),
+                    partitionKeyIndex,
+                    previous.getGetterName(),
+                    property.getGetterName());
```
I was wondering if this would generate the method multiple times when more than one concrete entity class is parsed, but it looks like `Messager` handles the de-duplication.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -434,26 +474,33 @@ private void scanFieldAnnotations(
-      Annotation annotation = getMethod.getAnnotation(annotationClass);
-      if (annotation != null) {
+      Optional<? extends ResolvedAnnotation<? extends Annotation>> annotation =
+          getMethodAnnotation(annotationClass, getMethod, typeHierarchy);
```
Nit: could you avoid static imports for non-obvious cases like this?

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/AnnotationScanner.java:
```diff
@@ -0,0 +1,74 @@
+          continue;
+        }
+
+        if (candidateMethod.getSimpleName().equals(getMethod.getSimpleName())) {
```
Comparing by name is enough because there's no way getters/setters with the same name but different types could coexist in the type hierarchy.
It also has a very interesting side-effect: the property type can be generic on a parent class/interface.
```java
interface Identifiable<IdentifierT> {
  @PartitionKey
  IdentifierT getId();
}
@Entity
public class Product implements Identifiable<UUID> { ... }
```

tolbertam committed:
Set highestAncestor to Object's TypeMirror by default

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java:
```diff
@@ -0,0 +1,241 @@
+        }
+        if (!atHighest) {
+          newInterfacesToScan.addAll(interfaceElement.getInterfaces());
+        }
```
1. If `atHighest` happens for an interface, we don't communicate that to the caller. Wouldn't it then continue to navigate superclasses?
2. If it happens for anything but the first interface, we've already added the parents of the previous interfaces at the same level. Example: assuming highest=I2, we will traverse I3 (and all its parents I think).
    ```
    I3
    |
    I1 I2
    |  /
     C
```

tolbertam replied:
> If atHighest happens for an interface, we don't communicate that to the caller. Wouldn't it then continue to navigate superclasses?

That's right, I had considered changing `highestAncestor` to `highestAncestors`, but after thinking about it some more realized that'd maybe be overthinking it.  Not sure we should expect very complex type hierarchies where you need to prevent scanning at both the class and interface level.  I could add that though if you think it is worth it.

> If it happens for anything but the first interface, we've already added the parents of the previous interfaces at the same level

Correct, that's the way I intended it.  If `I2` is highest, that effectively means don't scan anything above `I2`, since `I1` is a separate hierarchy, `I3` would also be scanned.

olim7t replied:
> Not sure we should expect very complex type hierarchies

Agreed, in practice it will probably be a couple of levels of either parent classes or interfaces.

If the current behavior is as intended, that's fine with me as long as we document it somewhere in the manual.

tolbertam replied:
I added details about this to `highestAncestor` javadoc in `HierarchyScanStrategy`.  I don't cover `HierarchyScanStrategy` in the manual, other than to mention and link to the javadoc for it in both inheritance sections as I expect a large percentage of the time the default behavior should be sufficient, and the javadoc does a good job explaining how to use it, but let me know if you'd like me to add more to the manual.

olim7t reviewed (changes requested):
Amazing how far you've taken this, I haven't yet found a case that isn't covered 👍 

See a bunch of remarks inline, most of them minor except for interface scanning, where I think we need to tighten a couple of edge cases.

tolbertam commented:
Thanks for the feedback @olim7t , working on addressing that all now :+1:

tolbertam committed:
Apply suggestions from code review

Co-Authored-By: Olivier Michallat <olim7t@users.noreply.github.com>

tolbertam committed:
Small feedback changes

tolbertam committed:
Process feedback

tolbertam commented:
I think I've addressed everything and this should be ready for another review pass :+1:

olim7t commented on mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/HierarchyScanStrategy.java:
```diff
@@ -0,0 +1,80 @@
+   * The {@link Class} to consider the highest ancestor, meaning the classes that this class extends
+   * or implements will not be scanned for annotations.
+   *
+   * <p>Note that a limitation of this configuration is the that the chosen highest ancestor only
```
```suggestion
   * <p>Note that a limitation of this configuration is that the chosen highest ancestor only
```

olim7t commented on mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/HierarchyScanStrategy.java:
```diff
@@ -0,0 +1,80 @@
+   * or implements will not be scanned for annotations.
+   *
+   * <p>Note that a limitation of this configuration is the that the chosen highest ancestor only
+   * applies to the part of the type hierarchy that contains this class.
```
I don't understand this what this paragraph means.

tolbertam replied:
yeah, this looks awkward and doesn't make sense at all.  It wasn't what I was trying to express at all, the following paragraph alone should be sufficient, so i'll just remove this.

olim7t reviewed (approved):
LGTM

tolbertam committed:
Update mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/HierarchyScanStrategy.java

Co-Authored-By: Olivier Michallat <olim7t@users.noreply.github.com>

tolbertam committed:
Cleanup documentation

tolbertam committed:
Fix small bug in EntityHelperGetMethodGenerator, tags should be property

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/EntityHelperGetMethodGenerator.java:
```diff
@@ -89,7 +89,7 @@ public EntityHelperGetMethodGenerator(
           // advantage in calling them instead of the generic get().
           getBuilder.addStatement(
-              "returnValue.$L(source.get($L, $T))",
+              "returnValue.$L(source.get($L, $L))",
```
Noticed tests failing on 2.1 and it appears because I missed `setTags` so `tags` was never considered a property in my tests.  On fixing this I noticed helpers failed to compile because this expected a type, but really it should be a literal since `enclosingClass.addGenericTypeConstant(typeName)` adds and returns a type constant name.

tolbertam commented:
think i've sorted out remaining issues, will merge if jenkins runs cleanly.
